compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala

```diff
@@ -978,36 +989,327 @@ class JSCodeGen()(using genCtx: Context) {
+    val sym = dd.symbol
+    val Block(stats, _) = dd.rhs
+    assert(sym.isPrimaryConstructor, s"called with non-primary ctor: $sym")
+
+    var jsSuperCall: Option[js.JSSuperConstructorCall] = None
+    val jsStats = List.newBuilder[js.Tree]
+
+    /* Move all statements after the super constructor call since JS
+     * cannot access `this` before the super constructor call.
+     *
+     * scalac inserts statements before the super constructor call for early
+     * initializers and param accessor initializers (including val's and var's
+     * declared in the params). We move those after the super constructor
+     * call, and are therefore executed later than for a Scala class.
+     */
+    withPerMethodBodyState(sym) {
+      stats.foreach {
+        case tree @ Apply(fun @ Select(Super(This(_), _), _), args)
+            if fun.symbol.isClassConstructor =>
+          assert(jsSuperCall.isEmpty, s"Found 2 JS Super calls at ${dd.sourcePos}")
+          implicit val pos: Position = tree.span
+          jsSuperCall = Some(js.JSSuperConstructorCall(genActualJSArgs(fun.symbol, args)))
+
+        case stat =>
+          val jsStat = genStat(stat)
+          assert(jsSuperCall.isDefined || !jsStat.isInstanceOf[js.VarDef],
+              "Trying to move a local VarDef after the super constructor call of a non-native JS class at " +
+              dd.sourcePos)
+          jsStats += jsStat
+      }
+    }
+
+    assert(jsSuperCall.isDefined,
+        s"Did not find Super call in primary JS construtor at ${dd.sourcePos}")
+
+    val params = dd.paramss.flatten.map(_.symbol)
+
+    new PrimaryJSCtor(sym, params, jsSuperCall.get :: jsStats.result())
+  }
+
+  private def genSecondaryJSClassCtor(dd: DefDef): SplitSecondaryJSCtor = {
+    val sym = dd.symbol
+    val Block(stats, _) = dd.rhs
+    assert(!sym.isPrimaryConstructor, s"called with primary ctor $sym")
+
+    val beforeThisCall = List.newBuilder[js.Tree]
+    var thisCall: Option[(Symbol, List[js.Tree])] = None
+    val afterThisCall = List.newBuilder[js.Tree]
+
+    withPerMethodBodyState(sym) {
+      stats.foreach {
+        case tree @ Apply(fun @ Select(This(_), _), args)
+            if fun.symbol.isClassConstructor =>
+          assert(thisCall.isEmpty,
+              s"duplicate this() call in secondary JS constructor at ${dd.sourcePos}")
+
+          implicit val pos: Position = tree.span
+          val sym = fun.symbol
+          thisCall = Some((sym, genActualArgs(sym, args)))
+
+        case stat =>
+          val jsStat = genStat(stat)
+          if (thisCall.isEmpty)
+            beforeThisCall += jsStat
+          else
+            afterThisCall += jsStat
+      }
+    }
+
+    val Some((targetCtor, ctorArgs)) = thisCall
+
+    val params = dd.paramss.flatten.map(_.symbol)
+
+    new SplitSecondaryJSCtor(sym, params, beforeThisCall.result(), targetCtor,
+        ctorArgs, afterThisCall.result())
+  }
+
+  private def genJSClassCtorDispatch(sym: Symbol, allParams: List[Symbol],
```

Comment:
The comments and code in this method appear to be somewhat different from the scala-js version, what are the high-level differences here?

Reply:
The main difference is that in dotc we don't have reliable param symbols. We can only work with the `paramNames` and `paramInfos` of a `MethodType`. This significantly complicates `jsParamInfos`, which cannot store param symbols, and cannot know whether params have a default value or not. For default values, we have to do some expensive checks to test whether a default getter exists or not. Since we only need that information at definition site of a method (but not at call site), we avoid the expensive checks in `jsParamInfos`, and instead we merge that information in here, by taking into account the param symbols found in the Trees.

Reply:
> The main difference is that in dotc we don't have reliable param symbols. 

Note that as of last year we do have https://github.com/lampepfl/dotty/blob/6922c1dd7a0be17cead711cdeb9f216327b4b368/compiler/src/dotty/tools/dotc/core/SymDenotations.scala#L300-L306, I don't know if it would help you here.